oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/MappedJwtClaimSetConverter.java

```diff
@@ -0,0 +1,288 @@
+		withDefaults.put(JwtClaimNames.ISS, new IssuerConverter());
+		withDefaults.put(JwtClaimNames.JTI, new StringConverter(JwtClaimNames.JTI));
+		withDefaults.put(JwtClaimNames.NBF, new TemporalConverter(JwtClaimNames.NBF));
+		withDefaults.put(JwtClaimNames.SUB, new StringConverter(JwtClaimNames.SUB));
+		withDefaults.putAll(attributeConverters);
+
+		MappedJwtClaimSetConverter converter = new MappedJwtClaimSetConverter(withDefaults);
+		issuedAtConverter.setAttributeConverters(converter.attributeConverters);
+
+		return converter;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public Map<String, Object> convert(Map<String, Object> claims) {
+		Assert.notNull(claims, "claims cannot be null");
+
+		Map<String, Object> mappedClaims = new HashMap<>(claims);
+
+		for (Map.Entry<String, Converter<Map<String, Object>, ?>> entry : this.attributeConverters.entrySet()) {
+			Converter<Map<String, Object>, ?> converter = this.attributeConverters.get(entry.getKey());
+			if (converter != null) {
+				put(mappedClaims, entry.getKey(), converter.convert(mappedClaims));
+			}
+		}
+
+		return mappedClaims;
+	}
+
+	private static class AudienceConverter implements Converter<Map<String, Object>, Collection<String>> {
+
+		@Override
+		public Collection<String> convert(Map<String, Object> source) {
+			Object audience = source.get(JwtClaimNames.AUD);
+
+			if (audience == null) {
+				return null;
+			}
+
+			if (audience instanceof Collection) {
+				return ((Collection<?>) audience).stream()
+						.filter(Objects::nonNull)
+						.map(Objects::toString)
+						.collect(Collectors.toList());
+			}
+
+			return Arrays.asList(audience.toString());
+		}
+	}
+
+	private static class IssuerConverter implements Converter<Map<String, Object>, URL> {
+
+		@Override
+		public URL convert(Map<String, Object> source) {
+			Object issuer = source.get(JwtClaimNames.ISS);
+
+			if (issuer == null) {
+				return null;
+			}
+
+			if (issuer instanceof URL) {
+				return (URL) issuer;
+			}
+
+			if (issuer instanceof URI) {
+				return toUrl((URI) issuer);
+			}
+
+			return toUrl(issuer.toString());
+		}
+
+		private URL toUrl(String issuer) {
+			try {
+				return new URL(issuer);
+			} catch (MalformedURLException e) {
+				throw new IllegalStateException("Could not coerce iss value to URL");
+			}
+		}
+
+		private URL toUrl(URI issuer) {
+			try {
+				return issuer.toURL();
+			} catch (MalformedURLException e) {
+				throw new IllegalStateException("Could not coerce iss value to URL");
+			}
+		}
+	}
+
+	private static class IssuedAtConverter extends TemporalConverter {
+
+		private Map<String, Converter<Map<String, Object>, ?>> attributeConverters;
+
+		public IssuedAtConverter() {
+			super(JwtClaimNames.IAT);
+		}
+
+		@Override
+		public Instant convert(Map<String, Object> source) {
+			Instant issuedAt = super.convert(source);
+			if (issuedAt != null) {
+				return issuedAt;
+			}
+
+			Instant expiresAt = getExpiresAt(source);
```

Comment:
I really view this as something distinct from conversion of the attributes. It is really defaulting the value. I think it might be best to place this in `MappedJwtClaimSetConverter` directly. This will also allow simplification of the API as suggested above

Reply:
I agree that this is more about defaults than conversion. I also think that we should focus just on conversion and not concern ourselves with defaulting properties in the first place. My motive, then, for placing this logic in the converter is to make it easy for users to override this defaulting.

Since this has the potential to change the contract (as discussed above), I'd like to hear your thoughts on other ways that users could easily override this defaulting of issuedAt. My concern is that I cannot do the fallback before conversion since it includes date arithmetic. And so, if it is after the set of conversions, then the user needs to completely reimplement the converter (or the JwtDecoder) in order to remove that defaulting.